### PR TITLE
Memleak

### LINKED ISSUE #135: miri doesn't drop values moved into closures with `let _ = x`
The following code works fine if `drop(x)` is used instead of `let _ = x`. The closure function generated contains just a `_0 = ()` and a `return`. Trans probably does some magic, but I haven't been able to locate it.

```rust
struct Foo<'a>(&'a mut bool);

impl<'a> Drop for Foo<'a> {
    fn drop(&mut self) {
        *self.0 = true;
    }
}

fn f<T: FnOnce()>(t: T) {
    t()
}

fn main() {
    let mut ran_drop = false;
    {
        let x = Foo(&mut ran_drop);
        let x = move || { let _ = x; };
        f(x);
    }
    assert!(ran_drop);
}
```

### PR DESCRIPTION
**Pull request by oli-obk**
check for memory leaks after the interpretation is finished and error if any memory leaks occurred.

After fixing the bugs leading to leaks, no regressions were noticed in the rustc test suite.

fixes #135

### CODE REVIEW

File: tests/compile-fail/memleak.rs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+//error-pattern: the evaluated program leaked memory
+
+fn main() {
+    std::mem::forget(Box::new(42));
```
Comment (eddyb) [inline tests/compile-fail/memleak.rs]: Could you also add a test for an `Rc` cycle? There should be examples around back from before `mem::forget` was made safe, with the whole scoped threads discussion.

Comment (oli-obk) [inline tests/compile-fail/memleak.rs]: done

File: tests/run-pass/move_fn_closure.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+struct Foo<'a>(&'a mut bool);
```
Comment (solson) [inline tests/run-pass/move_fn_closure.rs]: You should probably merge this file with the existing https://github.com/solson/miri/blob/master/tests/run-pass/closure-drop.rs.

File: tests/run-pass/closure-drop.rs [hunk 3]
```diff
@@ -26,5 +26,15 @@ fn main() {
         f(g);
```
Comment (solson) [inline tests/run-pass/closure-drop.rs]: I meant removing the `FIXME` above this and simplifying it all. :P

Comment (oli-obk) [inline tests/run-pass/closure-drop.rs]: gotta rustup first

Comment (oli-obk) [inline tests/run-pass/closure-drop.rs]: rustupped and fixed

**Pull request merged**